builtin/common/vector.lua

```diff
@@ -366,3 +368,22 @@ function vector.dir_to_rotation(forward, up)
 	end
 	return rot
 end
+
+if rawget(_G, "core") and core.set_read_vector and core.set_push_vector then
+	local function read_vector(v)
+		return v.x, v.y, v.z
+	end
+	core.set_read_vector(read_vector)
+	core.set_read_vector = nil
+
+	if rawget(_G, "jit") then
+		-- This is necessary to prevent trace aborts.
+		local function push_vector(x, y, z)
+			return (fast_new(x, y, z))
+		end
+		core.set_push_vector(push_vector)
+	else
+		core.set_push_vector(fast_new)
+	end
```

Comment:
For simplicity, you could just do the same in PUC lua as with luajit.

(Also unneeded parentheses around the `fast_new` call.)

Reply:
Without the parentheses, LuaJIT always aborts the trace, for some reason.

I chose to call `fast_new` directly with PUC Lua as a small optimization. It isn't much, but vectors may be created very frequently by C++, e.g. due to ABM calls.

Reply:
Interesting that the parentheses make a difference. Maybe it's because it's equivalent to `select(fast_new(...), 1)` and therefore breaks tail-calls? :thinking: 

Hm, yes the if doesn't hurt. And I haven't considered how often is called due to ABMs. It's fine then.